controllers/result.go

```diff
@@ -133,64 +134,57 @@ func GetResult(ctx context.Context, w http.ResponseWriter, r *http.Request) (int
 		return http.StatusUnauthorized, nil
 	}
 
-	if p.UserKey.Parent() != nil {
-		json.NewEncoder(w).Encode(result.Key(resultKey))
-		return http.StatusOK, nil
+	if p.UserKey.Parent() != nil && !util.HasParent(resultKey, p.UserKey) {
+		return http.StatusUnauthorized, nil
 	}
 
-	if !util.HasParent(resultKey, p.UserKey) {
-		return http.StatusUnauthorized, nil
+	if result.Finished == (time.Time{}) {
+		return createFinalResult(ctx, w, resultKey, result)
 	}
-	return createFinalResult(ctx, w, resultKey, result)
+
+	json.NewEncoder(w).Encode(result.Key(resultKey))
+	return http.StatusOK, nil
 }
 
-func createFinalResult(ctx context.Context, w http.ResponseWriter, resultKey *datastore.Key, result model.Result) (int, error) {
+func createFinalResult(ctx context.Context, w http.ResponseWriter, resultKey *datastore.Key, result model.Result) (status int, err error) {
 	go computeFinalScore(ctx, result)
 
 	var challenge model.Challenge
 	if err := datastore.Get(ctx, result.Challenge, &challenge); err != nil {
-		return http.StatusInternalServerError, nil
+		return http.StatusInternalServerError, err
 	}
 
-	var taskKey *datastore.Key
-	for _, taskKey = range challenge.Tasks {
-		switch taskKey.Kind() {
```

Comment:
Well, you have [`model.NewQueryForCodeSubmission`](https://github.com/coduno/api/pull/50/files#diff-89d9c4766c982b0a257b196029ca3625R175) there, which sets a kind..